rust/blockstore/src/arrow/block/delta/builder_storage.rs

```diff
@@ -0,0 +1,161 @@
+        self.storage.shrink_to_fit();
+        Self { storage: split_off }
+    }
+
+    fn len(&self) -> usize {
+        self.storage.len()
+    }
+
+    fn iter<'referred_data>(
+        &'referred_data self,
+    ) -> Box<dyn Iterator<Item = (&'referred_data CompositeKey, &'referred_data V)> + 'referred_data>
+    {
+        Box::new(self.storage.iter().map(|(k, v)| (k, v)))
+    }
+
+    fn into_iter(self) -> impl Iterator<Item = (CompositeKey, V)> {
+        self.storage.into_iter()
+    }
+}
+
+impl<V: ArrowWriteableValue> Default for VecBuilderStorage<V> {
+    fn default() -> Self {
+        Self {
+            storage: Vec::new(),
+        }
+    }
+}
+
+pub enum BuilderStorage<V: ArrowWriteableValue> {
+    BTreeBuilderStorage(BTreeBuilderStorage<V>),
+    VecBuilderStorage(VecBuilderStorage<V>),
+}
+
+impl<V: ArrowWriteableValue> BuilderStorage<V> {
+    pub fn add(&mut self, key: CompositeKey, value: V) {
+        match self {
+            BuilderStorage::BTreeBuilderStorage(storage) => storage.add(key, value),
+            BuilderStorage::VecBuilderStorage(storage) => storage.add(key, value),
+        }
+    }
+
+    pub fn delete(&mut self, key: &CompositeKey) -> Option<V> {
+        match self {
+            BuilderStorage::BTreeBuilderStorage(storage) => storage.delete(key),
+            BuilderStorage::VecBuilderStorage(storage) => storage.delete(key),
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        match self {
+            BuilderStorage::BTreeBuilderStorage(storage) => storage.len(),
+            BuilderStorage::VecBuilderStorage(storage) => storage.len(),
+        }
+    }
+
+    pub fn min_key(&self) -> Option<&CompositeKey> {
+        match self {
+            BuilderStorage::BTreeBuilderStorage(storage) => storage.min_key(),
+            BuilderStorage::VecBuilderStorage(storage) => storage.min_key(),
+        }
+    }
+
+    pub fn split_off(&mut self, key: &CompositeKey) -> Self {
+        match self {
+            BuilderStorage::BTreeBuilderStorage(storage) => {
+                BuilderStorage::BTreeBuilderStorage(storage.split_off(key))
+            }
+            BuilderStorage::VecBuilderStorage(storage) => {
+                BuilderStorage::VecBuilderStorage(storage.split_off(key))
+            }
+        }
+    }
+
+    pub fn iter<'referred_data>(
+        &'referred_data self,
+    ) -> Box<dyn Iterator<Item = (&'referred_data CompositeKey, &'referred_data V)> + 'referred_data>
+    {
+        match self {
+            BuilderStorage::BTreeBuilderStorage(storage) => storage.iter(),
+            BuilderStorage::VecBuilderStorage(storage) => storage.iter(),
+        }
+    }
+
+    pub fn into_iter(self) -> impl Iterator<Item = (CompositeKey, V)> {
+        match self {
+            BuilderStorage::BTreeBuilderStorage(storage) => Either::Left(storage.into_iter()),
+            BuilderStorage::VecBuilderStorage(storage) => Either::Right(storage.into_iter()),
```

Comment:
because these return `impl Iterator...` the return type of the match arms don't match; the `Either` wrapper fixes this

Reply:
either was already in our dependency tree but yeah it's not too bad to do ourselves